File: Source/code/forms/formOptimise.pas
// ===================================================================
//
// (c) Paul Alan Freshney 2012-2022
// www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
// https://github.com/MaximumOctopus/LEDMatrixStudio
//
// Please do not modifiy this comment section
//
// ===================================================================


unit formOptimise;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, ExtCtrls, Buttons,

  languagehandler,

  exportoptions,

  optimisation, xglobal;


type
  TfrmOptimise = class(TForm)
    Bevel1: TBevel;
    sbOptimise: TSpeedButton;
    mMemo: TMemo;
    sbCopyOutput: TSpeedButton;
    mData: TMemo;
    Panel1: TPanel;
    gbOutputOptions: TGroupBox;
    Label4: TLabel;
    Label5: TLabel;
    Label7: TLabel;
    Label6: TLabel;
    cbDataSize: TComboBox;
    cbLanguageFormat: TComboBox;
    cbPerRow: TComboBox;
    procedure sbOptimiseClick(Sender: TObject);
    procedure sbCopyOutputClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    MatrixData : TStringList;

    procedure SetGUILanguageText;
  public
    { Public declarations }
    MaxFrames : integer;
    HexFormat : boolean;
    ColumnsLSB, RowsLSB : integer;
    ExportFormat : integer;
    ColumnsDirection : integer;
    XType : integer;
    beforeData, afterData : integer;

    procedure PopulateMatrixData;

    function  GetDataSize: TNumberSize;
  end;

var
  frmOptimise: TfrmOptimise;

function DoOptimise: boolean;


implementation


uses utility;

{$R *.dfm}


function DoOptimise: boolean;
begin
  with TfrmOptimise.Create(Application) do
    try
      ShowModal;

      MatrixData.Free;

      Result := true;
    finally
      Free;
    end;
end;


procedure TfrmOptimise.FormCreate(Sender: TObject);
begin
  MatrixData := TStringList.Create;

  cbDataSize.Items.Add('4 bits (one nybble)');
  cbDataSize.Items.Add('1 byte');
  cbDataSize.Items.Add('2 bytes');
  cbDataSize.Items.Add('4 byte');
  cbDataSize.Items.Add('8 bytes');

  cbDataSize.ItemIndex := 1;

  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportCommaSeparated]);
  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportPICAXEEEPROM]);
  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportCCpp1Dimensional]);
  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportCCpp2Dimensional]);
  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportCCppFastLED]);
  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportPython1Dimensional]);
  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportPython2Dimensional]);
  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportMicrochip]);
  cbLanguageFormat.Items.Add(GLanguageHandler.Text[kExportPascal]);

  cbLanguageFormat.ItemIndex := 0;

  cbPerRow.Items.Add('4');
  cbPerRow.Items.Add('5');
  cbPerRow.Items.Add('7');
  cbPerRow.Items.Add('8');
  cbPerRow.Items.Add('10');
  cbPerRow.Items.Add('16');
  cbPerRow.Items.Add('20');
  cbPerRow.Items.Add('32');
  cbPerRow.Items.Add('64');

  cbPerRow.ItemIndex := 4;

  SetGUILanguageText;
end;


procedure TfrmOptimise.SetGUILanguageText;
begin
  Caption := GLanguageHandler.Text[kOptimise] + ' (beta)';

  sbOptimise.Caption := GLanguageHandler.Text[kOptimise];
  sbCopyOutput.Caption := GLanguageHandler.Text[kCopyOutput];

  gbOutputOptions.Caption := GLanguageHandler.Text[kOutputOptions];
  Label4.Caption := GLanguageHandler.Text[kDataSize];
  Label5.Caption := GLanguageHandler.Text[kLanguage];
  Label6.Caption := GLanguageHandler.Text[kOutput];
  Label7.Caption := GLanguageHandler.Text[kPerRow];
end;


procedure TfrmOptimise.sbCopyOutputClick(Sender: TObject);
 begin
  mMemo.CopyToClipboard;
end;


function TfrmOptimise.GetDataSize: TNumberSize;
begin
  case cbDataSize.ItemIndex of
    0 : Result := ns8Bit;
    1 : Result := ns8Bit;
    2 : Result := ns16bit;
    3 : Result := ns32bit;
    4 : Result := ns64bit;
  else
    Result := ns8Bit;
  end;
end;


procedure TfrmOptimise.sbOptimiseClick(Sender: TObject);
 var
  lOutput : TStringList;
  t : integer;
  teo : TExportOptions;

 begin
  lOutput := TStringList.Create;

  PopulateMatrixData;

  HexFormat  := True;
  ColumnsLSB := 0;
  RowsLSB    := 0;

  // ===========================================================================
  // ===========================================================================
  // ===========================================================================

  teo.NumberSize := GetDataSize;
  teo.Language   := TExportLanguage(cbLanguageFormat.ItemIndex);
  teo.LineCount  := StrToIntDef(cbPerRow.Text, 10);

  if OptimiseDataSimple(teo, MatrixData, lOutput) then begin

    mMemo.Clear;

    for t:= 0 to lOutput.Count - 1 do begin
      mMemo.Lines.Add(lOutput[t]);
    end;

  end;

  // ===========================================================================
  // ===========================================================================
  // ===========================================================================

  lOutput.Free;
end;


procedure TfrmOptimise.PopulateMatrixData;
var
  t, l : integer;
  lData : string;
  lLine : string;
  lInComment : boolean;

begin
  MatrixData.Clear;

  for t:= 0 to mData.Lines.Count - 1 do begin

    lLine := mData.Lines[t];

    lData      := '';
    lInComment := False;

    if (lLine <> '') then begin

      for l := 1 to length(lLine) do begin

        if (not(lInComment)) then begin

          if lLine[l] = ',' then begin
            MatrixData.Add(lData);

            lData := '';
          end
          else begin
            if (lLine[l] = '''') or
               (lLine[l] = '/') or
               (lLine[l] = ';') then
              lInComment := true
            else begin
              if TUtility.IsAlphaNumeric(lLine[l]) then
                lData := lData + lLine[l];
            end;
          end;
        end;

      end;

      if lData <> '' then
        MatrixData.Add(lData);
    end;
  end;
end;


end.

File: Source/Forms/FormOptimise.cpp
// ===================================================================
//
//   (c) Paul Alan Freshney 2012-2024
//   www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
//   https://github.com/MaximumOctopus/LEDMatrixStudio
//
//   https://maximumoctopus.hashnode.dev/
//
//   C++ Rewrite October 11th 2023
//
// ===================================================================

#include <vcl.h>
#pragma hdrstop

#include "FormOptimise.h"
#include "LanguageConstants.h"
#include "LanguageHandler.h"
#include "Optimisation.h"
#include "Utility.h"

extern LanguageHandler *GLanguageHandler;

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmOptimise *frmOptimise;

//---------------------------------------------------------------------------

void OpenOptimise(TheMatrix *tm)
{
	TfrmOptimise *frmOptimise = new TfrmOptimise(Application);

    frmOptimise->thematrix = tm;

	if (frmOptimise->ShowModal() == mrOk)
	{
	}

	delete frmOptimise;
}

//---------------------------------------------------------------------------

__fastcall TfrmOptimise::TfrmOptimise(TComponent* Owner)
	: TForm(Owner)
{
	cbDataSize->Items->Add(L"4 bits (one nybble)");
	cbDataSize->Items->Add(L"1 byte");
	cbDataSize->Items->Add(L"2 bytes");
	cbDataSize->Items->Add(L"4 byte");
	cbDataSize->Items->Add(L"8 bytes");

	cbDataSize->ItemIndex = 1;

	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportCommaSeparated].c_str());
	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportPICAXEEEPROM].c_str());
	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportCCpp1Dimensional].c_str());
	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportCCpp2Dimensional].c_str());
	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportCCppFastLED].c_str());
	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportPython1Dimensional].c_str());
	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportPython2Dimensional].c_str());
	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportMicrochip].c_str());
	cbLanguageFormat->Items->Add(GLanguageHandler->Text[kExportPascal].c_str());

	cbLanguageFormat->ItemIndex = 0;

	cbPerRow->Items->Add(L"4");
	cbPerRow->Items->Add(L"5");
	cbPerRow->Items->Add(L"7");
	cbPerRow->Items->Add(L"8");
	cbPerRow->Items->Add(L"10");
	cbPerRow->Items->Add(L"16");
	cbPerRow->Items->Add(L"20");
	cbPerRow->Items->Add(L"32");
	cbPerRow->Items->Add(L"64");

	cbPerRow->ItemIndex = 4;

	SetGuiLanguageText();
}


void __fastcall TfrmOptimise::sbCopyOutputClick(TObject *Sender)
{
	mMemo->CopyToClipboard();
}


void __fastcall TfrmOptimise::sbOptimiseClick(TObject *Sender)
{
	std::vector<std::wstring> output;

	PopulateMatrixData();

	HexFormat  = true;
	ColumnsLSB = 0;
	RowsLSB    = 0;

	// ===========================================================================

	ExportOptions teo;

	teo.Code.Size = GetDataSize();
	teo.Code.Language  = ExportLanguage(cbLanguageFormat->ItemIndex);
	teo.Code.LineCount = cbPerRow->Text.ToIntDef(10);

	mMemo->Clear();

	if (!Optimiser::OptimiseDataSimple(thematrix, teo, MatrixData, output))
	{
		mMemo->Lines->Add(L"Error :(");
	}

	for (int t = 0; t < output.size(); t++)
	{
		mMemo->Lines->Add(output[t].c_str());
	}
}


void TfrmOptimise::SetGuiLanguageText()
{
	std::wstring cap = GLanguageHandler->Text[kOptimise] + L" (beta)";

	Caption = cap.c_str();

	sbOptimise->Caption = GLanguageHandler->Text[kOptimise].c_str();
	sbCopyOutput->Caption = GLanguageHandler->Text[kCopyOutput].c_str();

	gbOutputOptions->Caption = GLanguageHandler->Text[kOutputOptions].c_str();
	Label4->Caption = GLanguageHandler->Text[kDataSize].c_str();
	Label5->Caption = GLanguageHandler->Text[kLanguage].c_str();
	Label6->Caption = GLanguageHandler->Text[kOutput].c_str();
	Label7->Caption = GLanguageHandler->Text[kPerRow].c_str();
}


void TfrmOptimise::PopulateMatrixData()
{
	MatrixData.clear();

	for (int t = 0; t < mData->Lines->Count; t++)
	{
		std::wstring line = mData->Lines->Strings[t].c_str();

		std::wstring data = L"";
		bool incomment = false;

		if (!line.empty())
		{
			for (int l = 0; l < line.length(); l++)
			{
				if (!incomment)
				{
					if (line[l] == L',')
					{
						MatrixData.push_back(data);

						data = L"";
					}
					else
					{
						if (line[l] == L'\'' ||
							line[l] == L'/' ||
							line[l] == L';')
						{
							incomment = true;
						}
						else
						{
							if (Utility::IsAlphaNumeric(line[l]))
							{
								data += line[l];
							}
						}
					}
				}
			}

			if (!data.empty())
			{
				MatrixData.push_back(data);
			}
		}
	}
}


NumberSize TfrmOptimise::GetDataSize()
{
	switch (cbDataSize->ItemIndex)
	{
	case 0:
		return NumberSize::k8Bit;
	case 1:
		return NumberSize::k8Bit;
	case 2:
		return NumberSize::k16bit;
	case 3:
		return NumberSize::k32bit;
	case 4:
		return NumberSize::k64bit;

	default:
		return NumberSize::k8Bit;
	}
}
